### PR TITLE
k8s: Disallow unknown fields in CNP & CCNP

### PR DESCRIPTION
**Pull request by christarazi**
See commit msgs.

Reported-by: André Martins <andre@cilium.io>

### CODE REVIEW

**Comment by christarazi** [comment]
test-me-please

**Review by aanm — APPROVED** [review]
Please add the Fixes <commit sha> to the commit, after the tests have passed

**Comment by aanm** [comment]
test-missed-k8s

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

File: pkg/k8s/apis/cilium.io/v2/types_test.go [hunk 1]
```diff
@@ -348,6 +348,26 @@ func (s *CiliumV2Suite) TestParseSpec(c *C) {
...
+
+	empty := &CiliumNetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "rule1",
+			UID:       uuidRule,
+		},
+	}
+	_, err = empty.Parse()
+	c.Assert(err.Error(), Matches, ".+empty policy")
```
Comment (christarazi) [inline pkg/k8s/apis/cilium.io/v2/types_test.go]: Update this to use the sentinel error (`ErrEmptyCNP`)

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

**Comment by christarazi** [comment]
test-me-please

File: pkg/k8s/apis/cilium.io/v2/ccnp_types.go [hunk 2]
```diff
@@ -91,14 +92,14 @@ func (r *CiliumClusterwideNetworkPolicy) Parse() (api.Rules, error) {
 
 	retRules := api.Rules{}
 
-	if r.Spec != nil {
+	switch {
```
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/ccnp_types.go]: This switch is changing behavior, with this switch a policy with both spec and specs defiled wouldn't return any errors. We should keep this functionality but move the `if r.Spec == nil && r.Specs == nil` to line 94.

File: pkg/k8s/apis/cilium.io/v2/cnp_types.go [hunk 3]
```diff
@@ -234,7 +235,8 @@ func (r *CiliumNetworkPolicy) Parse() (api.Rules, error) {
 
 	retRules := api.Rules{}
 
-	if r.Spec != nil {
+	switch {
+	case r.Spec != nil:
```
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/cnp_types.go]: Same comment left for the CCNP.

File: pkg/k8s/watchers/cilium_network_policy.go [hunk 4]
```diff
@@ -265,10 +266,20 @@ func (k *K8sWatcher) updateCiliumNetworkPolicyV2(ciliumNPClient clientset.Interf
...
+		ns := oldRuleCpy.GetNamespace() // Disambiguates CNP & CCNP
+
+		// We want to ignore parsing errors for empty policies, otherwise the
+		// update to the new policy will be skipped.
+		switch {
+		case ns != "" && !errors.Is(err, cilium_v2.ErrEmptyCNP):
+			fallthrough
+		case ns == "" && !errors.Is(err, cilium_v2.ErrEmptyCCNP):
+			log.WithError(err).WithField(logfields.Object, logfields.Repr(oldRuleCpy)).
+				Warn("Error parsing old CiliumNetworkPolicy rule")
```
Comment (aanm) [inline pkg/k8s/watchers/cilium_network_policy.go:276]: to be precise this should be CiliumClusterwideNetworkPolicy :-)

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/k8s/apis/cilium.io/v2/ccnp_types.go (hunk 2 above)
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/ccnp_types.go]: Good catch, will change.

**Comment by christarazi** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Pull request merged**